exonum/src/node/consensus.rs

```diff
@@ -154,99 +145,93 @@ impl NodeHandler {
                 msg,
                 *block.prev_hash(),
                 self.last_block_hash()
-            ));
+            );
         }
 
         if self.state.incomplete_block().is_some() {
-            return Err(format!(
-                "Already there is an incomplete block, msg={:?}",
-                msg
-            ));
+            bail!("Already there is an incomplete block, msg={:?}", msg);
         }
 
         if !msg.verify_tx_hash() {
-            return Err(format!("Received block has invalid tx_hash, msg={:?}", msg));
-        }
-
-        if let Err(err) = self.verify_precommits(&msg.precommits(), &block_hash, block.height()) {
-            return Err(format!("{}, block={:?}", err, msg));
+            bail!("Received block has invalid tx_hash, msg={:?}", msg);
         }
+        let precommits: Result<Vec<_>, _> = msg.precommits()
+            .into_iter()
+            .map(Precommit::verify_precommit)
+            .collect();
+        self.verify_precommits(&precommits?, &block_hash, block.height())?;
 
         Ok(())
     }
 
     /// Handles the `Block` message. For details see the message documentation.
     // TODO: Write helper function which returns Result. (ECR-123)
-    pub fn handle_block(&mut self, msg: &BlockResponse) {
-        if let Err(err) = self.validate_block_response(msg) {
-            error!("{}", err);
-            return;
-        }
+    pub fn handle_block(&mut self, msg: Message<BlockResponse>) -> Result<(), failure::Error> {
+        self.validate_block_response(&msg)?;
 
         let block = msg.block();
         let block_hash = block.hash();
         if self.state.block(&block_hash).is_none() {
             let snapshot = self.blockchain.snapshot();
             let schema = Schema::new(snapshot);
             let has_unknown_txs = self.state
-                .create_incomplete_block(msg, &schema.transactions(), &schema.transactions_pool())
+                .create_incomplete_block(&msg, &schema.transactions(), &schema.transactions_pool())
                 .has_unknown_txs();
 
             let known_nodes = self.remove_request(&RequestData::Block(block.height()));
 
             if has_unknown_txs {
                 trace!("REQUEST TRANSACTIONS");
-                self.request(RequestData::BlockTransactions, *msg.from());
+                self.request(RequestData::BlockTransactions, msg.author());
 
                 for node in known_nodes {
                     self.request(RequestData::BlockTransactions, node);
                 }
             } else {
-                self.handle_full_block(msg);
+                self.handle_full_block(&msg)?;
             }
         } else {
-            self.commit(block_hash, msg.precommits().iter(), None);
+            let precommits: Result<Vec<_>, _> = msg.precommits()
+                .into_iter()
+                .map(Precommit::verify_precommit)
+                .collect();
```

Comment:
Can we avoid using a temporary vector and use the map iterator instead?

Here and in other places below.

Reply:
We can, but then we need to slightly refactor code that work with Iterator.

Because `Precommit::verify_precommit` can fail, so collecting it to `Result<Vec<_>, _>` is imporatant to filter out errors.

Reply:
Heh... TIL that `collect()` treats `Result` in a special way. Thanks!

(This is what happens when you don't read the docs cover to cover and rely on your intuition.)